core/shared/src/main/scala/org/http4s/Service.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+
+import cats.Alternative
+import cats.Applicative
+import cats.Monad
+import cats.StackSafeMonad
+import cats.effect.kernel.Resource
+import cats.mtl.Local
+
+import Service._
+
+sealed abstract class Service[F[_], A, B] { self =>
```

Comment:
I think the following could improve readability: 

```suggestion
sealed abstract class Service[A, F[_], B] { self =>
```

This puts the parameters in the same order as they appear in the underlying function `A => F[B]` or `A => Resource[F, B]` type.

This is something I would have also wished for Kleisli.

Reply:
That messes partial unification up

Reply:
> That messes partial unification up

Is this a known problem in Scala 2.13, that partial unification would fail selectively over the first vs the second parameter? Does it still happens in Scala 3?

Speaking about the type system, would it be possible to make `A` contravariant and `B` covariant?

Reply:
A was contravariant until the last commit and could be again with some additional work.

Covariance of B runs into https://typelevel.org/blog/2018/09/29/monad-transformer-variance.html, though fs2 somehow made it work with similar signatures, so, maybe.

Reply:
> s this a known problem in Scala 2.13, that partial unification would fail selectively over the first vs the second parameter? Does it still happens in Scala 3?

it's not a "problem", it's literally the definition of partial unification. It's like expecting that you can have `foo: String => Int => Boolean` and do `foo(3)`